internal/crypto/src/raw_signature/rust_native/signers/rsa_signer.rs

```diff
@@ -0,0 +1,181 @@
+
+        let oid = &pki.algorithm.oid;
+        if !(oid == &RSA_OID || oid == &RSA_PSS_OID) {
+            return Err(RawSignerError::InvalidSigningCredentials(format!(
+                "unsupported private key algorithm ({oid})"
+            )));
+        }
+
+        let pkcs1_key = pkcs1::RsaPrivateKey::try_from(pki.private_key)
+            .map_err(|e| RawSignerError::InvalidSigningCredentials(e.to_string()))?;
+
+        if pkcs1_key.version() != pkcs1::Version::TwoPrime {
+            return Err(RawSignerError::InvalidSigningCredentials(
+                "multi-prime RSA keys not supported".to_string(),
+            ));
+        }
+
+        let n = BigUint::from_bytes_be(pkcs1_key.modulus.as_bytes());
+        let e = BigUint::from_bytes_be(pkcs1_key.public_exponent.as_bytes());
+        let d = BigUint::from_bytes_be(pkcs1_key.private_exponent.as_bytes());
+        let prime1 = BigUint::from_bytes_be(pkcs1_key.prime1.as_bytes());
+        let prime2 = BigUint::from_bytes_be(pkcs1_key.prime2.as_bytes());
+        let primes = vec![prime1, prime2];
+        let private_key = RsaPrivateKey::from_components(n, e, d, primes)
+            .map_err(|e| RawSignerError::InvalidSigningCredentials(e.to_string()))?;
+
+        let alg: RsaSigningAlg = match alg {
+            SigningAlg::Ps256 => RsaSigningAlg::Ps256,
+            SigningAlg::Ps384 => RsaSigningAlg::Ps384,
+            SigningAlg::Ps512 => RsaSigningAlg::Ps512,
+            _ => {
+                return Err(RawSignerError::InternalError(
+                    "RsaSigner should be used only for SigningAlg::Ps***".to_string(),
+                ));
+            }
+        };
+
+        Ok(RsaSigner {
+            alg,
+            cert_chain,
+            private_key,
+            cert_chain_len,
+            time_stamp_service_url,
+            time_stamp_size: 10000,
+            // TO DO: Call out to time stamp service to get actual time stamp and use that size?
+        })
+    }
+}
+
+impl RawSigner for RsaSigner {
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>, RawSignerError> {
+        let mut rng = rand::thread_rng();
+
+        match self.alg {
+            RsaSigningAlg::Ps256 => {
+                let s = SigningKey::<Sha256>::new(self.private_key.clone());
+                let sig = s.sign_with_rng(&mut rng, data);
+                Ok(sig.to_bytes().to_vec())
+            }
+
+            RsaSigningAlg::Ps384 => {
+                let s = SigningKey::<Sha384>::new(self.private_key.clone());
+                let sig = s.sign_with_rng(&mut rng, data);
+                Ok(sig.to_bytes().to_vec())
+            }
+
+            RsaSigningAlg::Ps512 => {
+                let s = SigningKey::<Sha512>::new(self.private_key.clone());
+                let sig = s.sign_with_rng(&mut rng, data);
+                Ok(sig.to_bytes().to_vec())
+            }
+        }
+    }
+
+    fn reserve_size(&self) -> usize {
+        1024 + self.cert_chain_len + self.time_stamp_size
```

Comment:
This should probably include OCSP size if part of this signer

Reply:
@mauricefisher64 do we have a way to estimate OCSP size ahead of time?